File: compiler/src/lexer.rs
use crate::elements::{ConstValue, DataType, Keyword, Token};


pub struct Lexer {
    input: String,
    pos: usize,
}

impl Lexer {
    pub fn new(input: String) -> Self {
        Lexer { input: input, pos: 0 }
    }
    
    pub fn tokenize(&mut self) -> Vec<Token> {
        let mut tokens: Vec<Token> = Vec::new();

        while let Some(token) = self.next_token() {
            tokens.push(token);
        }

        tokens
    }

    fn next_token(&mut self) -> Option<Token> {
        self.skip_unreadable();
        if self.pos >= self.input.len() {
            return None;
        }
        let current_char = self.input.chars().nth(self.pos)?;
        match current_char {
            // Кавычка - читаем строку
            '\"' => {
                self.pos += 1;
                let res: Option<Token> = Some(Token::Data { value: ConstValue::Str { value: self.read_string() } });
                self.pos += 1;
                return res;
            }
            // всякие символы и встроенные функции
            '(' => { self.pos += 1; Some(Token::LBr) }
            ')' => { self.pos += 1; Some(Token::RBr) }
            ',' => { self.pos += 1; Some(Token::Comma) }
            ':' => { self.pos += 1; Some(Token::Colon) }
            '+' => { self.pos += 1; Some(Token::Plus) }
            '-' => { self.pos += 1; Some(Token::Minus) }
            '*' => { self.pos += 1; Some(Token::Multiply) }
            '/' => { self.pos += 1; Some(Token::Divide) }
            '=' => { self.pos += 1; Some(Token::Equal) }
            // Цифра - читаем число
            _ if current_char.is_digit(10) => {
                return self.read_digit();
            }
            // Буква - читаем слово
            _ => {
                let word: String = self.read_word();

                self.check_word(&word[..])
            }
        }
    }

    fn read_string(&mut self) -> String {
        let start = self.pos;
        while let Some(c) = self.input.chars().nth(self.pos) {
            if c == '\"' {
                break;
            }
            self.pos += 1
        }

        let res: &String = &self.input[start..self.pos].to_string();
        
        res.to_string()

    }

    fn read_digit(&mut self) -> Option<Token> {
        let start = self.pos;
        let mut is_float = false;
        while let Some(c) = self.input.chars().nth(self.pos) {
            if !c.is_digit(10) && c != '.' { break; }
            if c == '.' { is_float = true; }
            self.pos += 1;
        }

        let num_str = &self.input[start..self.pos];

        if is_float {
            Some(Token::Data { value: ConstValue::Float { value: num_str.parse().ok()? } })
        }
        else {
            Some(Token::Data { value: ConstValue::Int { value: num_str.parse().ok()? } })
        }

    }

    fn read_word(&mut self) -> String {
        let start = self.pos;
        while let Some(c) = self.input.chars().nth(self.pos) {
            if !c.is_ascii_alphabetic() { break; }
            self.pos += 1;
        }

        let word: &String = &self.input[start..self.pos].to_string();

        word.to_string()
    }

    // Если не ключевое слово/встроенная функция - то идентификатор
    fn check_word(&mut self, word: &str) -> Option<Token> {
        let mut token: Option<Token>;
        
        token = self.check_keywords(word);
        if token != None {
            return token;
        }

        token = self.check_funcs(word);
        if token != None {
            return token;
        }

        token = self.check_datatypes(word);
        if token != None {
            return token;
        }

        Some(Token::Identifier { value: word.to_string() })
    }

    fn check_keywords(&mut self, word: &str) -> Option<Token> {
        match word {
            "fn" => { Some(Token::Keyword { value: Keyword::Fn }) }
            "returning" => { Some(Token::Keyword { value: Keyword::Returning }) }
            "return" => { Some(Token::Keyword { value: Keyword::Return }) }
            "body" => { Some(Token::Keyword { value: Keyword::Body }) }
            "args" => { Some(Token::Keyword { value: Keyword::Args }) }
            _ => { None }
        }
    }

    fn check_funcs(&mut self, word: &str) -> Option<Token> {
        match word {
            "+" => { Some(Token::Plus) }
            "-" => { Some(Token::Minus) }
            "*" => { Some(Token::Multiply) }
            "/" => { Some(Token::Divide) }
            "=" => { Some(Token::Equal) }
            "print" => { Some(Token::Print) }
            _ => { None }
        }
    }

    fn check_datatypes(&mut self, word: &str) -> Option<Token> {
        match word {
            "int" => { Some(Token::DataType { value: DataType::Int }) }
            "float" => { Some(Token::DataType { value: DataType::Float }) }
            "str" => { Some(Token::DataType { value: DataType::Str }) }
            "bool" => { Some(Token::DataType { value: DataType::Bool }) }
            "void" => { Some(Token::DataType { value: DataType::Void }) }
            _ => { None }
        }
    }

    fn skip_unreadable(&mut self) {
        while let Some(c) = self.input.chars().nth(self.pos) {
            if !(c == ' ' || c == '\n' || c == '\t') { break; }
            self.pos += 1;
        }
    }

}

File: C++_implementation/Flang/Compiler/Lexer.cpp
#include "pch.h"
#include "Lexer.h"
#include <cctype>

using namespace std;


Tokens::IToken* Lexer::nextToken() {
	skipUnreadable();
	char current = input[pos];

	switch (current)
	{
	case '\"':
	{
		pos++;
		Tokens::ConstValue* token = new Tokens::ConstValue(readString());
		pos++;
		return token;
	}

	case '(':
	{
		pos++;
		Tokens::LBr* token = new Tokens::LBr{};
		return token;
	}
	case ')':
	{
		pos++;
		Tokens::RBr* token = new Tokens::RBr{};
		return token;
	}
	case ',':
	{
		pos++;
		Tokens::Comma* token = new Tokens::Comma{};
		return token;
	}
	case ':':
	{
		pos++;
		Tokens::Colon* token = new Tokens::Colon{};
		return token;
	}
	default:
	{
		if (isdigit(current)) {
			string value = readDigit();
			Tokens::ConstValue* token = new Tokens::ConstValue(value);
			return token;
		}
		string word = readWord();
		return checkWord(word);
	}
	}
}

string Lexer::readString() {
	size_t start = pos;
	char current = input[start];
	while (pos < inputLength && current != '\"') {
		pos++;
		current = input[pos];
	}
	return input.substr(start, pos - start);
}

string Lexer::readDigit() {
	size_t start = pos;
	char current = input[start];
	while (pos < inputLength && (isdigit(current) || current == '.')) {
		pos++;
		current = input[pos];
	}
	return input.substr(start, pos - start);
}

string Lexer::readWord() {
	size_t start = pos;
	char current = input[start];
	while (pos < inputLength && (isalpha(current) || current == '_')) {
		pos++;
		current = input[pos];
	}
	return input.substr(start, pos - start);
}

Tokens::IToken* Lexer::checkWord(string word) {
	if (word == "defn") {
		Tokens::Keyword* token = new Tokens::Keyword(new Tokens::Defn{});
		return token;
	}
	else if (word == "return") {
		Tokens::Keyword* token = new Tokens::Keyword(new Tokens::Return{});
		return token;
	}
	else if (word == "int") {
		Tokens::DataType* token = new Tokens::DataType(new DataTypes::Int{});
		return token;
	}
	else if (word == "float") {
		Tokens::DataType* token = new Tokens::DataType(new DataTypes::Float{});
		return token;
	}
	else if (word == "str") {
		Tokens::DataType* token = new Tokens::DataType(new DataTypes::Str{});
		return token;
	}
	else if (word == "bool") {
		Tokens::DataType* token = new Tokens::DataType(new DataTypes::Bool{});
		return token;
	}
	else if (word == "void") {
		Tokens::DataType* token = new Tokens::DataType(new DataTypes::Void{});
		return token;
	}

	else {
		Tokens::Identifier* token = new Tokens::Identifier(word);
		return token;
	}
}

void Lexer::skipUnreadable() {
	while (input[pos] == ' ' || input[pos] == '\n' || input[pos] == '\t') {
		pos++;
	}
}

Lexer::Lexer(string input) {
	this->input = input;
	this->inputLength = input.length();
	this->pos = 0;
}

vector<Tokens::IToken*> Lexer::tokenize() {
	vector<Tokens::IToken*> result;
	while (pos < inputLength) {
		result.push_back(nextToken());
	}
	return result;
}